contracts/debt/TokemakIncurDebt.sol

```diff
@@ -0,0 +1,305 @@
+        require(!borrower[_account].isBorrower, "already borrower");
+        borrower[_account].isBorrower = true;
+    }
+
+    /**
+     * @notice sets the maximum debt limit for the system
+     * - onlyOwner (or governance)
+     * - limit must be greater than or equal to borrower's outstanding debt
+     * @param _account the address that will interact with contract
+     * @param _limit borrower's debt limit in OHM
+     */
+    function setBorrowerDebtLimit(address _account, uint256 _limit) external onlyGovernor {
+        require(borrower[_account].isBorrower, "not a borrower");
+        require(_limit >= borrower[_account].debt);
+        require(_limit < globalDebtLimit, "_limit above global debt limit");
+
+        borrower[_account].limit = _limit;
+    }
+
+    /**
+     * @notice sets the maximum debt limit for the system
+     * - onlyOwner (or governance)
+     * - borrower must not have outstanding debt
+     * @param _account the address that will interact with contract
+     */
+    function revokeBorrower(address _account) external onlyGovernor {
+        require(borrower[_account].debt == 0, "_account still has outstanding debt");
+        borrower[_account].isBorrower = false;
+    }
+
+    /**
+     * @notice deposits gOHM/sOHM to use as collateral
+     * - msg.sender must be a borrower
+     * - this contract must have been approved _amount
+     * @dev will unwrap and hold as sOHM in this contract
+     * @param _amount amount of gOHM
+     */
+    function deposit(uint256 _amount, address _token) external {
+        require(borrower[msg.sender].isBorrower, "not a borrower");
+        IERC20(_token).safeTransferFrom(msg.sender, address(this), _amount);
+
+        if (_token == gOHM) {
+            uint256 sbalance = IStaking(staking).unwrap(address(this), _amount);
+
+            borrower[msg.sender].collateralInSOHM += sbalance;
+            borrower[msg.sender].collateralInGOHM += _amount;
+        } else if (_token == sOHM) {
+            uint256 gbalance = IgOHM(gOHM).balanceTo(_amount);
+
+            borrower[msg.sender].collateralInSOHM += _amount;
+            borrower[msg.sender].collateralInGOHM += gbalance;
+        }
+    }
+
+    /**
+     * @notice allow borrowers to borrow OHM
+     * - msg.sender must be a borrower
+     * - _ohmAmount must be <= availableDebt
+     * @param _ohmAmount amount of OHM to borrow
+     */
+    function borrow(uint256 _ohmAmount) external {
+        require(borrower[msg.sender].isBorrower, "not a borrower");
+        require(
+            _ohmAmount <= borrower[msg.sender].limit - borrower[msg.sender].debt,
+            "_ohmAmount more than borrower limit"
+        );
+        require(_ohmAmount <= getAvailableToBorrow(), "_ohmAmount more than available loan");
+
+        ITreasury(treasury).incurDebt(_ohmAmount, OHM);
+        IERC20(OHM).safeTransfer(msg.sender, _ohmAmount);
+
+        borrower[msg.sender].debt += _ohmAmount;
```

Comment:
I think we can move this above the interactions, to stick to the check-effects-interactions pattern.

Reply:
Adapted